sklearn/cluster/hierarchical.py

```diff
@@ -444,6 +447,39 @@ def linkage_tree(X, connectivity=None, n_components=None,
         assert n_clusters <= n_samples
         n_nodes = 2 * n_samples - n_clusters
 
+    if linkage == 'single':
+        from scipy.sparse.csgraph import minimum_spanning_tree
+
+        # Use scipy.sparse.csgraph to generate a minimum spanning tree
+        mst = minimum_spanning_tree(connectivity.tocsr())
+
+        # Convert the graph to scipy cluster array format
+        nonzeros = mst.nonzero()
+        nonzero_vals = mst.data
+        mst_array = np.vstack(nonzeros + (nonzero_vals,)).T
+
+        # Sort edges of the min_spanning_tree by weight
+        mst_array = mst_array[np.argsort(mst_array.T[2]),:]
+
+        # Convert edge list into standard hierarchical clustering format
+        single_linkage_tree = _hierarchical.single_linkage_label(mst_array)
```

Comment:
This converts the MST into scipy.cluster.hierarchy format. It just uses a union find so is pretty fast, and I already had code that does this (and was well tested), so it seemed the most efficient way to get something working. I am open to other options.